active/0021-transfer-files-over-mqtt.md

```diff
@@ -0,0 +1,199 @@
+
+## Requirements
+
+* The protocol MUST use only PUBLISH type of MQTT Control Packet
+* The protocol MUST support transfer of file segments
+* Server MUST be able to verify integrity of each file segment and of the whole file
+* Client MAY know total file size when initiating the transfer
+* Client MAY abort file transfer
+* Server MAY ask the client to pause file transfer
+* Server MAY ask the client to abort file transfer
+* The protocol MUST guarantee "At least once" delivery
+* Server MUST NOT support subscription on topics dedicated for file transfer
+
+## AWS IoT MQTT-based file delivery (reference design)
+
+As an example of existing implementation we can look at AWS IoT Core [which provides functionality](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery.html) to [deliver files to IoT devices](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery-in-devices.html):
+
+![](0021-assets/aws-mqtt-file-delivery.png)
+
+## Design
+
+### Overview
+
+* Files are split in segments, segments can be of arbitrary length
+* Client MUST generate UUID according to [RFC 4122](https://www.rfc-editor.org/rfc/rfc4122) for each file being transferred and use it as file Id in Topic Name
+* Broker MUST validate that provided UUID conforms to [RFC 4122](https://www.rfc-editor.org/rfc/rfc4122)
+* Client calculates sha256 checksum of the segment it's about to send and sends it as part of Topic Name
+* Broker MUST calculate checksum of the segment data it received and verify that it matches the one in Topic Name
+* Client uses $file Topic to transfer files
+* Broker MUST NOT let clients subscribe to $file topics
+* Segment length can be calculated on the server side by subtracting the length of the [Variable Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901025) from the [Remaining Length](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901105) field that is in the [Fixed Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901021)
+* Data is transferred in PUBLISH packets in the following order:
+  1. $file/{fileId}/init
+  2. $file/{fileId}/{offset}/{sha256sum}
+  3. $file/{fileId}/{offset}/{sha256sum}
+  4. ...
+  3. $file/{fileId}/[fin/{sha256sum} | abort]
+
+#### `$file/{fileId}/init` message
+
+Initialize the file transfer. Server is expected to store metadata from the payload in the session along with `{fileId}` as a reference for the rest of file metadata.
+
+  * Qos=1
+  * DUP=0 for the initial segment transfer, 1 when retransmitting
+  * Payload Format Indicator=0x01
+  * `{fileId}` is corresponding file UUID
+  * Payload is a JSON document
+
+##### `init` payload JSON Schema
+
+Available [here](https://github.com/emqx/mqtt-file-transfer-schema/blob/v1.0.0/init.json).
+
+* Broker SHOULD have default setting for `segments_ttl`
+* Broker MAY delete segments of unfinished file transfers when their TTL has expired
```

Comment:
We should probably add an optional field to hold the assembled sha256 checksum.
And make the per-segment checksum optional too.